spec.md

```diff
@@ -96,7 +96,63 @@ and eventually have it accepted as a Unicode Technical Standard (UTS).
 
 ## Data Model
 
-> _The data-only representation of a message._
+A Message provides the representation of a single message.
+It takes one of two forms,
+either as a PatternMessage or a SelectMessage.
+In a multi-locale context,
+a single message's choice of representation as one or the other
+should not affect its representation in other languages or locales.
+
+```ts
+type Message = PatternMessage | SelectMessage
+```
+
+A PatternMessage contains a list of PatternElement values,
+some of which are directly defined literal values,
+while others are placeholders with formatted values that depend on additional data.
+
+```ts
+interface PatternMessage {
+  type: 'message'
+  value: PatternElement[]
+  meta?: Meta
+}
+```
+
+SelectMessage provides for the selection of one list of PatternElement values
+to use as the message's value when formatting,
+depending on the value of one or more Selector values.
+
+Each SelectCase is defined by a `key` of one or more string identifiers,
+and selection between them is made according to the corresponding Selector values.
+From this it follows that a valid SelectMessage must have at least as many `select` entries
+as its highest count of SelectCase `key` entries within its `cases`.
+The `fallback` value of a Selector is used in addition to its `value`
+when selecting one of the `cases` during formatting.
+It should match exactly one of the corresponding SelectCase `key` values.
+
+As a SelectMessage is not a valid PatternElement,
+it is not possible for it to be placed within the `value` of another message.
+
+```ts
+interface SelectMessage {
+  type: 'select'
+  select: Selector[]
+  cases: SelectCase[]
```

Comment:
Unfortunately that doesn't really work, because the `SelectCase` `key` is a `string[]`, and using an array type as a key would make the data model incompatible with JSON as well most other data representation formats. YAML might be the only one that would work.

Reply:
I understand.

But what about put in the spec what makes most sense from functionality perspective, and worry about syntax / implementation later?
What we need like an "abstract data type" feels like a map, no?
- the number of items in the 2 arrays are the same
- it's not ok to have doubles in the "select" array

It does not even supposed to compile
(although I have this a map in my ts implementation, and compiles, and works)

Reply:
I think we are looking at things from a level that is not abstract enough, is too close to (some) implementations.

I think here we should talk at the level of "Abstract Data Types"
https://en.wikipedia.org/wiki/Abstract_data_type

At that level you say "Stack" or "Queue" and people understand what that means (FIFO vs LILO).
You say "Map" (or "Associative Array") that means certain behavior.

We lower the semantic level, and then we need to add it back as documented behavior
"the two arrays should be of equal size" and "no duplicate keys" and "how to fallback  / recover in case of duplicate keys"

It's like saying "a box with chairs and wheels to move people around" instead of Car.

The tendency is always increase the semantic content: we define types, sometimes just aliases to existing ones.
We defines css styles so that we don't deal with collections of attributes.

Reply:
Transformed into a `Map<string[], PatternMessage>` in the abstract data type definition.

Reply:
@eemeli  question : we are duplicating interfaces to keep the JSON canonical representation?

Reply:
@romulocintra I would have preferred not needing to define interfaces separately as "abstract data" and their JSON compatible form, but as @mihnita is insisting on using a `Map` rather than an array + dropping the `type` values, this is necessary.